src/index.ts

```diff
@@ -0,0 +1,241 @@
+}
+
+interface OpeningParenthesis {
+  type: 'OpeningParenthesis';
+  value: '(';
+}
+
+type NumberOrBinaryOperationToken = NumberToken | BinaryOperationToken;
+
+type Token =
+  | BinaryOperationToken
+  | NumberToken
+  | OpeningParenthesis
+  | ClosingParenthesis;
+
+const OPERATORS = new Set(['+', '-', '*', '/']);
+
+const operatorPrecedence = {
+  '*': 2,
+  '/': 2,
+  '+': 1,
+  '-': 1,
+};
+
+const VALID_CHARACTERS = new Set(['+', '/', '-', '*', '.', '(', ')']);
+
+for (let i = 0; i < 10; i++) {
+  VALID_CHARACTERS.add(String(i));
+}
+
+const isOperator = (element: string) => OPERATORS.has(element);
+const isNumber = (element: string) => /^-?\d+$/.test(element);
+
+const hasConsecutiveOperators = (string: string): boolean => {
+  for (let i = 0; i < string.length; i++) {
+    const current = string[i];
+    const next = string[i + 1];
+
+    if (isOperator(current) && isOperator(next) && next !== '-') {
+      return true;
+    }
+  }
+
+  return false;
+};
+
+const eliminateSpaces = (string: string): string => string.replace(/\s/g, '');
+
+const hasValidCharacters = (string: string): boolean =>
+  string.split('').every((character) => VALID_CHARACTERS.has(character));
+
+const generateTokenForExpression = (string: string): Array<Token> => {
+  const trimmedString = eliminateSpaces(string);
+
+  if (!hasValidCharacters(trimmedString)) throw 'Invalid input';
+  if (hasConsecutiveOperators(trimmedString)) throw 'Syntax error';
+
+  const tokenArray = [];
+
+  let numberAccumulator = '';
+  let numberIsNegative = false;
+
+  const generateNumberToken = (value: string): NumberToken => ({
+    type: 'NumberToken',
+    value: numberIsNegative
+      ? Number(numberAccumulator) * -1
+      : Number(numberAccumulator),
+  });
+
+  const addNumberToken = (token: NumberToken) => {
+    tokenArray.push(token);
+    numberAccumulator = '';
+    numberIsNegative = false;
+  };
+
+  for (let i = 0; i < trimmedString.length; i++) {
+    const current = trimmedString[i];
+    const previous = trimmedString[i - 1];
+
+    if (isNumber(current) || current === '.') {
+      numberAccumulator += current;
+    } else if (current === '-' && i === 0) {
+      numberIsNegative = true;
+    } else if (isOperator(current)) {
+      if (numberAccumulator !== '') {
+        const token = generateNumberToken(numberAccumulator);
+        addNumberToken(token);
+      }
+      if (current === '-' && isOperator(previous)) {
+        numberIsNegative = true;
+      } else {
+        tokenArray.push({
+          type: 'BinaryOperationToken',
+          operation: current,
+        });
+      }
+    } else if (current === '(') {
+      tokenArray.push({
+        type: 'OpeningParenthesis',
+        value: '(',
+      });
+    } else if (isNumber(current) || current === ')') {
+      const token = generateNumberToken(numberAccumulator);
+      addNumberToken(token);
+      tokenArray.push({
+        type: 'ClosingParenthesis',
+        value: ')',
+      });
+    }
+  }
```

Comment:
documentation can especially be useful for a long chain of `if...else` like this at the top of the branching logic

Something along the lines of 

/*
IF ...
ELSE IF ...
ELSE ...
*/ 

in plain english